src/dynam.jl

```diff
@@ -0,0 +1,207 @@
+using Catlab
+using Catlab.CategoricalAlgebra
+using Catlab.Theories
+import Catlab.Theories: id, dom, codom, compose
+
+"""    VectorField{F}
+
+A dynamical system over a Euclidean Space on `n` dimensions. 
+Use `VectorField{Function}` to disable specialization on the vector field function.
+"""
+struct VectorField{F}
+    domain::FinSet # Rⁿ
+    field::F  # maps Rⁿ to Tangents(Rⁿ)
+end
+
+domain(U::VectorField) = U.domain
+tanspace(U::VectorField) = U.domain
+vfield(U::VectorField) = U.field
+
+(U::VectorField)(u) = begin
+    FinSet(length(u)) == domain(U) || error("Domain Mismatch: state vector has length $(length(u)) was expecting $(domain(U))")
+    vfield(U)(u)
+end
+
+"""    VectorFieldHom
+
+The structure for storing a homomorphism between VectorField objects. The key axiom is that
+
+    restrict(f, simulate(f, u) - f.u(u)) == 0:domain(f.v)
+
+this means that you can see a version of v inside u by applying the `Pullback(f)` map to restrict states of u to a state of v.
+Then apply the vector field for v to get a tangent vector for v. Then you can apply the `Pushforward(f)` map to get a tangent vector for u.
+
+    v::VectorField
+    u::VectorField
+    f::FinFunction
+   
+    Pullback(f):    domain(u) → domain(v)
+    Pushforward(f): tan(v) → tan(u)
+
+"""
+struct VectorFieldHom
+    v::VectorField
+    u::VectorField
+    f::FinFunction
+end
+
+dom(ϕ::VectorFieldHom) = ϕ.v
+codom(ϕ::VectorFieldHom) = ϕ.u
+
+id(V::VectorField) = VectorFieldHom(V, id(domain(V)))
+compose(ϕ::VectorFieldHom, γ::VectorFieldHom) = begin
+    VectorFieldHom(dom(ϕ),
+                codom(γ),
+                ϕ.f ⋅ γ.f)
+end
+
+pullback(f::FinFunction) = u -> u[collect(f)]
+pushforward(f::FinFunction) = u̇ -> map(codom(f)) do i
+    sum(u̇[j] for j in preimage(f, i);init=0.0)
+end
+
+""" Pullback{F} wraps F in a struct to make a callable for the action of pulling a function back along `f::F`. 
+For `F <: FinFunction`, f: N → M sends Xᴹ to Xᴺ by precomposition.
+"""
+struct Pullback{F}
+    f::F
+end
+
+(fꜛ::Pullback{F})(u) where F <: FinFunction = u[collect(fꜛ.f)]
```

Comment:
Cursed that sub/superscript `+` and `-` are available but not `*`. Unicode strikes again.

Reply:
Yeah, unicode just needs _ and ^ to be diacritic characters that combine with the next character.